sqlwhat/test_exercise.py

```diff
@@ -21,23 +23,33 @@ def test_exercise(
     """
     """
 
-    state = State(
-        student_code=student_code,
-        solution_code=solution_code,
-        pre_exercise_code=pre_exercise_code,
-        student_conn=student_conn,
-        solution_conn=solution_conn,
-        student_result=student_result,
-        solution_result=solution_result,
-        reporter=Reporter(errors=error),
-        force_diagnose=force_diagnose,
-    )
-
-    SCT_CTX["Ex"].root_state = state
+    reporter = Reporter(errors=error)
 
     try:
-        exec(sct, SCT_CTX)
-    except TestFail as tf:
-        return tf.payload
+        state = State(
+            student_code=student_code,
+            solution_code=solution_code,
+            pre_exercise_code=pre_exercise_code,
+            student_conn=student_conn,
+            solution_conn=solution_conn,
+            student_result=student_result,
+            solution_result=solution_result,
+            reporter=reporter,
+            force_diagnose=force_diagnose,
+        )
+
+        # the available SCT methods
+        sct_dict = get_checks_dict(checks)
+
+        # the available global variables
+        sct_context = create_sct_context(sct_dict, state)
+
+        exec(sct, sct_context)
+
+    except Failure as e:
+        if isinstance(e, InstructorError):
+            # TODO: decide based on context
```

Comment:
This TODO could be here a while, maybe expand it a little?

Reply:
For example in the future we could make an InstructorError a TestFail, making it appear in the left hand side and doesn't crash your session.